[ZH] Implement Headless Mode

helmutbuhler opened the pull request:
This PR adds a commandline option -headless which runs the game without any graphics.
This is necessary to implement a silent replay checker (see issue #565). Note that I have that replay checker already implemented, but I will push that in another PR. This is just preparation for that.

Note that right now, if you pass in -headless, the game just runs the Shellmap. Since it's headless, you won't be able to see or do anything. But if you run it in Debug, you can enjoy looking at the log file getting bigger.

Headless mode is implemented here by essentially only running GameLogic, and not the GameClient. Unfortunately, GameLogic depends on the GameClient in some small hacky ways, which makes this a non-trivial task. See also this comment I added:
```
// TheSuperHackers @logic-client-separation helmutbuhler 04/11/2025
// Some information about the architecture and headless mode:
// The game is structurally separated into GameLogic and GameClient.
// The Logic is responsible for everything that affects the game mechanic and what is synchronized over
// the network. The Client is responsible for rendering, input, audio and similar stuff.
// 
// Unfortunately there are some places in the code that make the Logic depend on the Client.
// (Search for @logic-client-separation)
// That means if we want to run the game headless, we cannot just disable the Client. We need to disable
// the parts in the Client that don't work in headless mode and need to keep the parts that are needed
// to run the Logic.
// The following describes which parts we disable in headless mode:
//
//	GameEngine:
//		TheGameClient is partially disabled:
//			TheKeyboard = NULL
//			TheMouse = NULL
//			TheDisplay is partially disabled:
//				m_3DInterfaceScene = NULL
//				m_2DScene = NULL
//				m_3DScene = NULL
//				(m_assetManager remains!)
//			TheWindowManager = NULL
//			TheIMEManager = NULL
//			TheTerrainVisual is partially disabled:
//				TheTerrainTracksRenderObjClassSystem = NULL
//				TheW3DShadowManager = NULL
//				TheWaterRenderObj = NULL
//				TheSmudgeManager = NULL
//				TheTerrainRenderObject is partially disabled:
//					m_treeBuffer = NULL
//					m_propBuffer = NULL
//					m_bibBuffer = NULL
//					m_bridgeBuffer = NULL
//					m_waypointBuffer = NULL
//					m_roadBuffer = NULL
//					m_shroud = NULL
//		TheRadar = RadarDummy
```

As it is now, this PR sets all the client components as listed above to NULL and adds all the appropriate NULL-checks (which are sadly quite a few).
For TheRadar I added a dummy implementation to avoid the NULL-checks in the many callers (this doesn't make much sense in the other components though).

If we decide to fix these separation issues properly in the future, we will be able to revert many of the changes in this PR in the future.

Alternatively, we can also decide to just hide the window in headless mode and initialize DX without using it. Then almost all the changes in this PR become obsolete as well. The question is then whether this will still run in CI.

I would prefer to keep the changes in this PR and to clean up these separation issues properly long-term. Having logic and client properly separated can help in the future when we decide to increase the framerate of the client, or if we want to implement a headless server. It's also nice to automatically test that the logic works without the client.

## TODO

- [ ] Replicate in Generals

helmutbuhler commented on GeneralsMD/Code/GameEngine/Include/Common/Radar.h:
```diff
@@ -293,6 +293,30 @@ class Radar : public Snapshot,
+class RadarHeadless : public Radar
+{
+public:
+	RadarHeadless(){}
```
Ah, what was I thinking...

helmutbuhler commented:
@feliwir 
Thanks for the review! I implemented all suggestions, and also just added some changes I forgot earlier.
I also converted to draft because we are still waiting on PR #544

helmutbuhler commented on GeneralsMD/Code/GameEngine/Source/Common/RandomValue.cpp:
```diff
@@ -249,7 +249,7 @@ Int GetGameClientRandomValue( int lo, int hi, const char *file, int line )
 #ifdef DEBUG_RANDOM_CLIENT
 DEBUG_LOG(( "%d: GetGameClientRandomValue = %d (%d - %d), %s line %d\n",
-				TheGameLogic->getFrame(), rval, lo, hi, file, line ));
+				TheGameLogic ? TheGameLogic->getFrame() : -1, rval, lo, hi, file, line ));
```
Note, this has nothing to do with headless, we don't set TheGameLogic to null. But this can crash on startup when Logic hasn't been initialized yet and when DEBUG_RANDOM_CLIENT is defined. So I snuck a fix in here.

helmutbuhler commented:
Alright, PR #544 is merged and I updated this PR, it's ready for review now. Please read the first post before you review though.

helmutbuhler commented:
Quick update with merge conflict fix and an added previously missed null pointer check

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Diplomacy.cpp:
```diff
@@ -169,6 +169,8 @@ BriefingList* GetBriefingTextList(void)
 //-------------------------------------------------------------------------------------------------
 void UpdateDiplomacyBriefingText(AsciiString newText, Bool clear)
 {
+	if (TheWindowManager == NULL)
```
There are a lot of TheWindowManager null checks. Can we perhaps have a Null implementation for the window manager and avoid all these null checks?

xezon commented on GeneralsMD/Code/GameEngineDevice/Source/W3DDevice/GameLogic/W3DTerrainLogic.cpp:
```diff
@@ -358,7 +360,8 @@ Real W3DTerrainLogic::getLayerHeight( Real x, Real y, PathfindLayerEnum layer, C
 Bool W3DTerrainLogic::isCliffCell( Real x, Real y) const
 {
-
+	// TheSuperHackers @logic-client-separation helmutbuhler 04/11/2025
```
DD/MM/YYYY

helmutbuhler replied:
I don't think we should use that. Search for /2003 and you'll find tons of dates using M/D/YYYY.
In my opinion, we should remain consistent with that, use the ISO format YYYY-MM-DD, or no dates at all.

xezon replied:
This format was set in 2021. If it needs to be changed, then it needs to be refactored in both Patch and Code repository.

xezon commented on GeneralsMD/Code/GameEngineDevice/Source/W3DDevice/GameClient/W3DView.cpp:
```diff
@@ -1538,6 +1538,8 @@ void W3DView::draw( void )
 	CustomScenePassModes customScenePassMode  = SCENE_PASS_DEFAULT;
 	Bool preRenderResult = false;
 
+	if (TheGlobalData->m_headless)
```
Shouldn't this test be higher up near
```cpp
	if(TheGlobalData->m_playIntro || TheGlobalData->m_afterIntro)
	{
		// redraw all views, update the GUI
		{
			TheDisplay->DRAW();
		}
		{
			TheDisplay->UPDATE();
		}
		return;
	}
```

and

```cpp
	// update display
	{
		TheDisplay->UPDATE();
	}

	{
		USE_PERF_TIMER(GameClient_draw)
			
	// redraw all views, update the GUI
	//if(TheGameLogic->getFrame() >= 2)
		
		TheDisplay->DRAW();
	}
```
?

Same with W3DDisplay::draw

xezon commented on GeneralsMD/Code/GameEngineDevice/Source/W3DDevice/GameClient/W3DDisplay.cpp:
```diff
@@ -457,12 +457,15 @@ W3DDisplay::~W3DDisplay()
 	// shutdown
 	Debug_Statistics::Shutdown_Statistics();
-	W3DShaderManager::shutdown();
+	if (!TheGlobalData->m_headless)
```
This looks suspicious. I would expect W3DDisplay to not even be created in Headless. It looks to be necessary for asset manager and ww math.

helmutbuhler replied:
Yeah, it's unfortunate, but TerrainLogic uses the Asset Manager to get the Terrain data. When we move that data to the TerrainLogic, we'll be able to skip creation of W3DDisplay in headless mode and remove all those checks in that file.

xezon commented on GeneralsMD/Code/GameEngineDevice/Source/W3DDevice/GameClient/BaseHeightMap.cpp:
```diff
@@ -2368,7 +2374,7 @@ void BaseHeightMapRenderObjClass::staticLightingChanged( void )
 	m_curNumScorchVertices=0;
 	m_curNumScorchIndices=0;
-	m_roadBuffer->updateLighting();
+	if (m_roadBuffer) m_roadBuffer->updateLighting();
```
line breaks

several times in this change.

helmutbuhler replied:
I changed it, but according to Google Guidelines it's allowed for short statements. It's also more readable in my opinion.

xezon replied:
The practical issue with `if (condition) action;` is that it does not allow to put a breakpoint for `action` in the source code. Therefore this style will be banned.

xezon commented on GeneralsMD/Code/GameEngineDevice/Source/W3DDevice/GameClient/W3DAssetManager.cpp:
```diff
@@ -965,6 +970,8 @@ int W3DAssetManager::Recolor_Mesh(RenderObjClass *robj, const int color)
 int W3DAssetManager::Recolor_HLOD(RenderObjClass *robj, const int color)
 {
 	int didRecolor=0;
+	if (TheGlobalData->m_headless)
```
Can move higher up. Several times in this change.

xezon reviewed (commented):
I would like to see more null implementations where so possible. I suspect the null checks were only placed to avoid crashing, right? I suspect it still runs a lot of logic that is not necessary to run in headless?

helmutbuhler commented:
@xezon 
Thanks for the review! Hopefully I can implement those requested changes soon.
But in the meantime, maybe it would be a good idea to also test whether these changes allow the game to run in Github actions container without gpu support. Unfortunately I don't have much experience there, maybe someone else can do that?

xezon commented on GeneralsMD/Code/GameEngine/Include/GameClient/GameWindowManager.h:
```diff
@@ -380,6 +380,49 @@ extern WindowMsgHandledType PassMessagesToParentSystem( GameWindow *window,
 
 
+// TheSuperHackers @feature helmutbuhler 24/04/2025
+// GameWindow that does nothing. Used for Headless Mode.
```
I suggest to move these dummies into new header + source files because they are a bit on the larger side.

jaapdeheer replied:
It's 3 dummies: `RadarDummy`, `GameWindowDummy` & `GameWindowManagerDummy`.
I moved `GameWindowDummy` from `GameWindowManager.h` to `GameWindow.h` because that makes more sense to me.

So now we have:
- a tiny `RadarDummy` in `Radar.h`
- a tiny `GameWindowDummy` in `GameWindow.h`
- and a not-so-tiny, but not huge, `GameWindowManagerDummy` in `GameWindowManager.h` & `GameWindowManager.cpp`

We _could_ move all of those into their own separate files, but for `RadarDummy` and `GameWindowDummy` that seems overkill since they're just a few lines each.
We can't really put them all in one separate `.h` & `.cpp` file together, since `RadarDummy` is not part of `GameClient` but `GameWindowDummy` & `GameWindowManagerDummy` are. (We could put them all in `Common` somewhere but that seems wrong.)

So alternatively we could move _only_ `GameWindowManagerDummy` into its own `.h` & `.cpp` file, but that would then be inconsistent with `RadarDummy` and `GameWindowDummy` and thus vaguely confusing.

So, I would suggest to leave it like this. What do you think?

helmutbuhler replied:
I think xezon meant only to move GameWindowDummy and GameWindowManagerDummy into a separate file. But I think the change you did is fine as well, so I'll wait on what xezon has to say.

xezon replied:
Ok. Can leave as is.

xezon replied:
On second look, AudioManagerDummy looks quite big as well. It would be nice if these new dummy classes would sit in their own headers that that they are nicely put away, out of view.

helmutbuhler replied:
I'm confused, what do you want :)

xezon commented on GeneralsMD/Code/GameEngineDevice/Source/W3DDevice/GameClient/W3DView.cpp:
```diff
@@ -616,7 +616,7 @@ void W3DView::setCameraTransform( void )
 	m_3DCamera->Set_Transform( cameraTransform );
 
-	if (TheTerrainRenderObject) 
+	if (TheTerrainRenderObject && W3DDisplay::m_3DScene != NULL) 
```
Is a call into `W3DView::setCameraTransform` needed at all? Can this call not be prevented higher up?

helmutbuhler replied:
I looked into it. It doesn't really make sense because `W3DView::setCameraTransform` has so many callers. But I moved the check to the top of the function.

xezon commented on GeneralsMD/Code/GameEngineDevice/Source/W3DDevice/GameLogic/W3DGhostObject.cpp:
```diff
@@ -457,7 +458,7 @@ void W3DGhostObject::restoreParentObject(void)
 				//object that is in the scene, then it was probably added because the model
 				//changed while the object was ghosted (for damage states, garrison, etc.).
-				if (robj->Peek_Scene() == NULL)
+				if (robj->Peek_Scene() == NULL && W3DDisplay::m_3DScene != NULL)
```
Is a call to `W3DGhostObject::restoreParentObject` required? Perhaps this call can be prevented higher up?

helmutbuhler replied:
`W3DGhostObject` is a Logic Object, so it wouldn't make sense to move that check higher up the callstack. But I moved this check up inside the function as well.

xezon commented on GeneralsMD/Code/GameEngineDevice/Source/W3DDevice/GameClient/W3DTerrainVisual.cpp:
```diff
@@ -655,13 +658,17 @@ Bool W3DTerrainVisual::load( AsciiString filename )
 	}
 	// add our terrain render object to the scene
-	W3DDisplay::m_3DScene->Add_Render_Object( m_terrainRenderObject );
+	if (W3DDisplay::m_3DScene != NULL)
```
Looking at all these `m_3DScene` tests, I wonder if a few `m_headless` tests can be placed higher up the call stack or around these `m_3DScene` usages to avoid the bulk of rendering logic. Assuming that the rendering does not affect game logic, it should be possible to strip a lot of the render updates, no?

helmutbuhler replied:
For the final PR with the replay simulation, I won't call GameClient::update at all, so don't worry :) I just wanted to make this PR testable and added some checks in the rendering to make it work. I can remove those checks in the next PR.

xezon replied:
How do we keep track of these check? Add a comment behind them?

xezon replied:
> For the final PR with the replay simulation, I won't call GameClient::update at all, so don't worry :) I just wanted to make this PR testable and added some checks in the rendering to make it work. I can remove those checks in the next PR.

How will this look? Which null tests will be gone?

helmutbuhler replied:
Okay, I've looked into it. Unfortunately the checks in `W3DTerrainVisual::load` cannot really be simplified because the `m_terrainRenderObject->initHeightData` call is logic relevant.

xezon replied:
So most of these new `m_3DScene` null checks will not be avoided in the future?

helmutbuhler replied:
No, sorry.

xezon reviewed (commented):
I do wonder if more of the null checks can be avoided with early checks
```cpp
if (TheGlobalData->m_headless)
  return;
```
or
```cpp
if (!TheGlobalData->m_headless)
  doRendering();
```

The advantage of these checks is
1) avoid more unecessary computations
2) make our headless related checks easier to find

I am also fond of the dummy implementations.

jaapdeheer commented on GeneralsMD/Code/GameEngine/Include/GameClient/GameClient.h:
```diff
@@ -259,4 +259,44 @@ inline Drawable* GameClient::findDrawableByID( const DrawableID id )
+//					m_waypointBuffer = NULL
+//					m_roadBuffer = NULL
+//					m_shroud = NULL
+//		TheRadar = RadarDummy
```
Done as suggested, in 174ab672 & 8cdad718

helmutbuhler commented on GeneralsMD/Code/GameEngineDevice/Source/W3DDevice/GameLogic/W3DTerrainLogic.cpp:
```diff
@@ -281,6 +281,8 @@ Real W3DTerrainLogic::getGroundHeight( Real x, Real y, Coord3D* normal ) const
 {
 #define USE_THE_TERRAIN_OBJECT
 #ifdef USE_THE_TERRAIN_OBJECT
+	// TheSuperHackers @logic-client-separation helmutbuhler 04/11/2025
```
Including the Particle SW PR, these should be complete now.

helmutbuhler commented:
Alright, I just pushed another update.

I tested headless mode with about 1400 various replays and they all work as expected now. I had to fix a few more issues:
- mss32.dll was crashing in rare and random situations when multiple instances were running. I added AudioManagerDummy to fix this
- GameWindowManagerDummy had a bug and was causing some memory corruption. I fixed that.
- Maps with bridges depend on m_bridgeBuffer in TheTerrainRenderObject, so I include that now in headless mode without the attached vertexbuffers.

I also added MouseDummy to eliminate some more NULL-checks. All issues should be solved now. Thanks again to @jaapdeheer for helping me fix some of them :)

Caball009 commented on GeneralsMD/Code/GameEngine/Include/Common/GameAudio.h:
```diff
@@ -372,6 +372,57 @@ class AudioManager : public SubsystemInterface
+	virtual void setDeviceListenerPosition(void) {}
+};
+
+
```
_This comment also applies to some other functions in this pr._

Many of these functions have unused function parameters. This generates compiler warnings at high warning levels. It may be worth considering adding `[[maybe_unused]]`, though this will likely cause VC6 to trip up. Removing or commenting out the variable names would also get rid of the warnings, I think. 

Specifying a `void` 'function parameter' is only advisable for C and completely meaningless for C++ AFAIK.

xezon replied:
The `void` is simply a product of copy & paste original code.

This comment is unrelated to this change and affects whole code base.

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/GameClient.cpp:
```diff
@@ -535,7 +538,7 @@ void GameClient::update( void )
 	if(TheGlobalData->m_afterIntro && !TheDisplay->isMoviePlaying())
 	{
-		if( playSizzle && TheGlobalData->m_playSizzle )
+		if( playSizzle && TheGlobalData->m_playSizzle && !TheGlobalData->m_headless )
```
Perhaps just set `m_playSizzle` to false when `m_headless` is set to true?

helmutbuhler replied:
Will be gone in next PR

helmutbuhler commented on GeneralsMD/Code/GameEngine/Include/GameClient/Mouse.h:
```diff
@@ -371,7 +371,20 @@ class Mouse : public SubsystemInterface
+	virtual void setCursor(MouseCursor cursor) {}
+	virtual void capture() {}
+	virtual void releaseCapture() {}
+	virtual UnsignedByte getMouseEvent(MouseIO *result, Bool flush) { return MOUSE_NONE; }
```
Yes, `Mouse` has some small functionality that's required in headless mode to prevent crashes. (Some functions return pointers to members)

helmutbuhler commented:
> How will this look? Which null tests will be gone?
I just pushed an update where I commented the checks that will be gone in the Replay Simulation PR

xezon commented:
Do you need help replicating this to Generals?

helmutbuhler commented:
> Do you need help replicating this to Generals?

I got the diff with:
`git diff 291382d374e15fabd86b393324badaa79f9bf5ce..6bb77c0f09b49f73a4ee869d8d833c0ac54ffb8b > changes.patch`

But there are quite a few conflicts with Generals. (Doesn't surprise me)
So I suggest we do that in a separate PR

helmutbuhler commented:
But I can give it a try and will call for help when I get stuck

xezon commented:
I can help you with it.

xezon commented:
I think the "Merge main" commits make this really hard to replicate to Generals with that method.

helmutbuhler commented:
I can do it, I just want to wait until this is approved because I don't want to sync two PRs if more changes are requested.

The merge commits are not the problem, it's that there are differences between Gen and ZH where this PR does changes. You can get the diff with the merge base with:
```
git diff $(git merge-base main branch)..branch
```
or use the git command I posted above.

xezon commented on GeneralsMD/Code/GameEngineDevice/Source/W3DDevice/GameLogic/W3DGhostObject.cpp:
```diff
@@ -434,7 +436,7 @@ void W3DGhostObject::removeParentObject(void)
 {
 	Drawable *draw=m_parentObject->getDrawable();
-	if (!draw)
+	if (!draw || TheGlobalData->m_headless)
```
How about
```cpp
if (TheGlobalData->m_headless)
  return;
```
at the top instead?

xezon commented on GeneralsMD/Code/GameEngineDevice/Source/W3DDevice/GameClient/W3DDisplay.cpp:
```diff
@@ -2089,7 +2109,7 @@ void W3DDisplay::createLightPulse( const Coord3D *pos, const RGBColor *color,
 																	 )
 {
-	if (innerRadius+attenuationWidth<2.0*PATHFIND_CELL_SIZE_F + 1.0f) {
+	if (innerRadius+attenuationWidth<2.0*PATHFIND_CELL_SIZE_F + 1.0f || m_3DScene == NULL) {
```
Perhaps make this
```cpp
if (m_3DScene == NULL)
  return;
```
at the top instead.

helmutbuhler commented:
Finally merged, thanks everyone! PR for the changes in Generals is here: #899
